HTTP: add handlers to allow object creation and modification

mildred opened the pull request:
This is for issue #611

jbenet commented on core/corehttp/gateway.go:
```diff
@@ -6,11 +6,13 @@ import (
+			return err
+		}
+		mux.Handle("/ipfs/", gateway)
+		return nil
```
:+1: on this approach to options

jbenet commented on cmd/ipfs/daemon.go:
```diff
@@ -36,6 +37,7 @@ the daemon.
 	Options: []cmds.Option{
 		cmds.BoolOption(initOptionKwd, "Initialize IPFS with default settings if not already initialized"),
 		cmds.BoolOption(mountKwd, "Mounts IPFS to the filesystem"),
+		cmds.BoolOption(readOnlyKwd, "Make the HTTP gateway read only (disable POST, PUT and DELETE)"),
```
let's flip this-- i think the default should be read only and writing should need enabling (at least for the time being). maybe:

```
-w --writable "Enable writing objects (with POST, PUT, and DELETE)"
```

jbenet commented on core/corehttp/gateway_handler.go:
```diff
@@ -163,23 +194,170 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	key, err := i.node.DAG.Add(newnode)
+	if err != nil {
+		webError(w, "Could not add recursively new node", err, http.StatusInternalServerError)
```
nitpick: `"Could not recursively add new node"`

jbenet commented on core/corehttp/gateway_handler.go:
```diff
@@ -163,23 +194,170 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	var newnode *dag.Node
+	if path[len(path)-1] == '/' {
+		newnode = NewDagEmptyDir()
```
should there be an early return here? is it easy for this to get mishandled below if people modify the code?

mildred replied:
It's not really possible to make this condition an early return. The new node still need to be saved with its parent directories. And there is semantically no difference ahead in code. We have a new node, and the rest of the function is to update the parent nodes.

jbenet commented on core/corehttp/gateway_handler.go:
```diff
@@ -163,23 +194,170 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	http.Redirect(w, r, "/ipfs/"+key.String()+"/", http.StatusCreated)
+}
+
+func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request, path string) {
```
this function is pretty long and there's lots of edge-case handling. Let's add some comments like so:

```go
// resolving path components into merkledag nodes
// if no links are found, create empty directories -- which will be linked and populated below.
path_nodes, err := i.node.Resolver.ResolveLinks(rootnd, components[:len(components)-1])
if _, ok := err.(p.ErrNoLink); ok {
  // Create empty directories, links will be made further down the code
  for len(path_nodes) < len(components) {
    path_nodes = append(path_nodes, NewDagEmptyDir())
  }
} else if err != nil {
  webError(w, "Could not resolve parent object", err, http.StatusBadRequest)
  return
}
```

jbenet commented on merkledag/node.go:
```diff
@@ -134,6 +134,16 @@ func (n *Node) Copy() *Node {
+	err := newnode.RemoveNodeLink(name)
+	err = nil // ignore error
+	err = newnode.AddNodeLink(name, that)
+	return newnode, err
```
hmmm not sure if we want ipns to be mutable. I dont know this part of the codebase very well. I'll let @whyrusleeping shepherd the IPNS parts here

mildred replied:
That's why the Node is not mutated by the function. It is copied on a new instance which is modified. In fact, that was already done like that in another part of the code (`update()` function in ` fuse/ipns/ipns_unix.go`).

jbenet replied:
Ahhh yes, i had missed the `n.Copy()`

jbenet commented on path/path.go:
```diff
@@ -13,16 +13,25 @@ import (
-func (s *Resolver) ResolvePath(fpath string) (*merkledag.Node, error) {
+// SplitPath clean up and split fpath. It extracts the first component (which
+// must be a Multihash) and return it separately.
+func SplitPath(fpath string) (mh.Multihash, []string, error) {
```
Maybe `SplitAbsPath` will make the semantics clearer to readers

jbenet commented on path/path.go:
```diff
@@ -39,13 +48,36 @@ func (s *Resolver) ResolvePath(fpath string) (*merkledag.Node, error) {
+		return nil, err
+	} else {
+		return nodes[len(nodes)-1], err
+	}
```
always check for `err != nil` because the other thing may return `err` even with partially valid output (and in most cases we'll want to suppress valid outputs when there are errors). I get the intent to check for nil on the slice... maybe:
```
if err != nil || nodes ==  nil {
    return nil, err
}
return nodes[len(nodes)-1], nil
```

jbenet commented:
@mildred thanks for the PR! a few comments above:
- [ ] switch mode option `readable --> writable`
- [ ] more comments on the putHandler
- [ ] `SplitAbsPath`
- [ ] small nit picks

And, @mappum and @whyrusleeping can you CR the gateway + ipns parts? (i dont have all that state as fresh in my mind so you'll be able to decide correctness better)

mildred commented:
I am going to update the code to fix the small issues you reported. Thank you for taking time to review my code.

mildred commented:
I rebased the branch fixing the things you noticed.

jbenet commented:
Modifications LGTM-- @mappum @whyrusleeping your parts?

mappum commented:
LGTM, this will be a very cool API for building DAG objects. But there be some conflicts to resolve when rebasing, sorry.

BTW I'll probably add an API that uses this in my JS library.

briantigerchow commented on core/corehttp/gateway.go:
```diff
@@ -6,11 +6,13 @@ import (
+		if err != nil {
+			return err
+		}
+		mux.Handle("/ipfs/", gateway)
```
There's some complexity emerging in `ServeHTTP`. There may be a simpler way to handle multiple HTTP methods on the same path. If we switch over to the gorilla mux, we get the following...

```Go
state := &gatewayState{node: n, writable: writable, dir: load()}
ipfsPath := r.PathPrefix("/ipfs/").Subrouter()
ipfsPath.Methods("GET").HandlerFunc(getHandler(state))
ipfsPath.Methods("PUT").HandlerFunc(putHandler(state))
ipfsPath.Methods("POST").HandlerFunc(postHandler(state))
ipfsPath.Methods("DELETE").HandlerFunc(deleteHandler(state))
```

Thoughts?

jbenet replied:
I'm in favor of moving to `gorilla.Mux`

jbenet commented on core/corehttp/gateway_handler.go:
```diff
@@ -14,7 +15,9 @@ import (
 	"github.com/jbenet/go-ipfs/importer"
 	chunk "github.com/jbenet/go-ipfs/importer/chunk"
 	dag "github.com/jbenet/go-ipfs/merkledag"
+	p "github.com/jbenet/go-ipfs/path"
```
longer import name -- `p` is too unrecognizable for comfort

jbenet commented:
Hey @mildred the major thing on my end is tests -- I want to get IPFS to a git-like reliability level. Being careful to test everything is what will get us there. -- i worry that including this major piece of functionality without tests might regress quickly (there's lots of different cases in `putHandler` and `deleteHandler` to take into account. Are you familiar with the sharness tests? @mappum has started writing some with `curl` for the gateway in a branch. this might be good here.  If you need this in sooner, i'd be fine merging and adding those in a separate PR, but ideally would go in here. Thoughts?

whyrusleeping commented on path/path.go:
```diff
@@ -55,19 +87,22 @@ func (s *Resolver) ResolvePath(fpath string) (*merkledag.Node, error) {
 //
 // ResolveLinks(nd, []string{"foo", "bar", "baz"})
 // would retrieve "baz" in ("bar" in ("foo" in nd.Links).Links).Links
 func (s *Resolver) ResolveLinks(ndd *merkledag.Node, names []string) (
```
Why was this change necessary?

mildred replied:
We need this because when we PUT to /ipfs/HASH/dir/dir/file, the hash of file changes. This makes necessary to update the directories in succession up to the root hash. I felt it was better not to parse the path two times (once for getting the leaf file hash, one other time to rewrite the directories). Instead the path parsing is done once using the ResolveLinks function

jbenet replied:
@whyrusleeping if you want the signature of `ResolveLinks` not to change, we can move this functionality to a differently-named function.

mildred commented:
Thanks jbenet for your reply. Indeed, I think tests are a very good addition. Unfortunately, I'm not too skilled in them, but I'll have to learn.

Do you think tests that are run from `go test` are enough or should we really test from the outside using curl ? I'd go for the former, and I'll update my pull request. I'm currently learning to do tests in my Ed25519 PR (#677) and they are really useful and easy to do, once you know what to test.

jbenet commented:
@mildred the more tests the better. I think for something like this (an http api), we'll always want end-to-end tests (testing it as a black box). We can do those in Go, using the `http` package and doing Gets and Puts.  if we're careful about the tests written, go tests can be fine. etc. doing them in sharness makes it obvious that something really works outside of go + the same process, and ensures that the formats are all correct (sharness makes it very hard to trick yourself).

The sharness tests are easy to pick up, too. take a look at https://github.com/jbenet/go-ipfs/blob/master/test/sharness/t0080-repo.sh

mildred commented:
I updated the pull request. Removed complexity from the ServeHttp function, implemented /ipns GET requests (as this code was lying around me unused and was part of a simplification refactoring).

I did not switch to gorilla Mux, there seem to be different implementations around (one at github.com the other at code.google.com). And I think this might involve `gateway.go#GatewayOption` or `gateway_handler.go#newGatewayHandler` and switching away from the current Mux.

I still have to make some tests.

jbenet commented:
@mildred i think @mappum rebased your branch for you (we merged more stuff into this area). what branch @mappum ?

mappum commented:
I rebased here: https://github.com/jbenet/go-ipfs/commits/http-rest

The tests might need to go back to using curl so we can test PUT and DELETE, but otherwise it should be RFM.

jbenet commented:
> The tests might need to go back to using curl 

is curl available everywhere? (we could install it with make)

mildred commented:
I also updated this branch on my side: see https://github.com/mildred/go-ipfs/commits/http-rest-wip. Rebasing and adding sharness tests using curl and shell scripting (see https://github.com/mildred/go-ipfs/commit/3cb8dbd9d107b0f9dd2ce5b8e55a7e062bc20cd0). But I seem to get in conflict with pull request #674. So I'll take a little time to get it in working order.

mildred commented:
I rebased my pull request. I fixed up the code and added sharness tests

jbenet commented:
@mildred fantastic! :+1: :+1: :+1:  Let's get this merged in!

jbenet commented:
Hm travis failed with this: https://travis-ci.org/jbenet/go-ipfs/jobs/48904581  (does this need adding stuff to `cmd/ipfs-gateway-fs` ? ) 

And sorry for the rebasing-- result of high-velocity dev. Do you want contributor access? it may make working on these larger pieces easier. I.e. i could push a commit to your branch that fixes that test.

briantigerchow commented:
> Do you want contributor access?

:+1:

mildred commented:
Thank you. I would be honoured to get contributor access. The work model would be to work on a branch on this repository, and everything else as usual, right? I'm not sure that I'll be as available as I am these days, though. And I have a few other projects as well.

Thank you for pointing me the travis log. I wasn't understanding how to find that info in travis until just now. I just updated the code to fix the build issue.

mildred commented:
The fmt import was removed from cmd/ipfs/daemon.do in commit 73234479fa095f0a3fd01581ae1aee749d92f6bc and this broke the build. Re-updated the pull request (home it will pass this time)

jbenet commented:
> Thank you. I would be honoured to get contributor access. 

Great, you've got it. 

> The work model would be to work on a branch on this repository, and everything else as usual, right? 

Correct. the only rules are:
- dont push straight to master (unless I ok it) 
- dont mess up other people's refs (unless they ok it) 
(basically, each branch has an "owner", and i own `master`)

Other notes:
- rebase often to stay on top of master
- pick long-ish branch names to avoid clashes
- if you rebase other people's branches, leave a comment in the associated PR (and be in contact with them)
- careful deleting refs (i.e. branches, tags) -- make sure nobody's using them / wants them

> I'm not sure that I'll be as available as I am these days, though. And I have a few other projects as well.

No worries

jbenet commented:
I've a couple comments on the sharness tests, but I'll merge this now to avoid more divergence. rest can be a separate PR
